### ISSUE TITLE
Binpack: multi-line argument should not end up on the same line as another argument

### ISSUE THREAD

**Issue by sjrd** [olafurpg/scalafmt#502]
**Version**: 0.4.3
**Integration**: sbt
**Configuration**: (for example, `--maxColumn 120`)

Original:
```scala
@deprecated(
    "Use @JSExport with an explicit option bag instead. " +
    "See the Scaladoc for more details.",
    "0.6.11")
class JSExportNamed extends scala.annotation.StaticAnnotation {
  def this(name: String) = this()
}
```

Scalafmt:
```scala
@deprecated(
    "Use @JSExport with an explicit option bag instead. " +
    "See the Scaladoc for more details.", "0.6.11")
class JSExportNamed extends scala.annotation.StaticAnnotation {
  def this(name: String) = this()
}
```
I would like the input to look like the original.

Note how `"0.6.11"` ends up being on the same line as the second half of the first argument. This should not happen. If an argument is multi-line, it should keep its lines to itself, and not share them with preceding or following arguments.

This should probably be the default behavior for binpacking. I don't know how this works in non-binpacking.

**Comment by olafurpg** [comment]
I recall you had a rule that directly conflicts with this rule. The rule was that if an argument crossed multiple lines it should be on its own line, for example

```scala
// limit                 |
// ok
function(a, b, d,
    foobar(kaz, baz,
        banana),
    e, k)
// not ok
function(a, b, d,
    foobar(kaz, baz,
        banana), e, k)
```
Which one do you prefer?

I can't find the related commit or issue, I definitely remember we discussed this offline.

If you allowed 2 space indentation for `+` then this wouldn't be an issue ;)

**Comment by sjrd** [comment]
Hum, I think it's actually exactly the *same* rule. If an argument crosses multiple lines, it should be on its own line. Equivalently, it must not share any of its lines with any other argument.

**Comment by sjrd** [comment]
> If you allowed 2 space indentation for `+` then this wouldn't be an issue ;) 

I don't understand why. It think in this case I would get the following, which is also not OK:
```scala
@deprecated(
    "Use @JSExport with an explicit option bag instead. " +
      "See the Scaladoc for more details.", "0.6.11")
class JSExportNamed extends scala.annotation.StaticAnnotation {
  def this(name: String) = this()
}```

**Comment by olafurpg** [comment]
Whoops, I misunderstood your issue. I thought you wanted to have the arguments on the same line! Now it makes sense.

> If you allowed 2 space indentation for + then this wouldn't be an issue ;)

I thought you had issue with these two lines aligning:

```scala
    "See the Scaladoc for more details.",
    "0.6.11")
```

Nevermind me!

**Comment by olafurpg** [comment]
BTW. I haven't looked at bin packing at all since the last time we went over the diffs together. I agree the current implementation is far from polished.

Unfortunately, I can't promise when I'll have time to look at this. As much as I would love to get the Scala.js style production ready, I don't expect it will be in the near future since my free time is already quite constrained.

**Comment by sjrd** [comment]
That's fine. I'm just going through the diffs of `scala-js/scala-js/library/` and reporting everything I see there ;)

**Comment by olafurpg** [comment]
Absolutely! Please report your issues. You never know, maybe one quiet evening I'll sit down with a nice warm cup of tea and feel super excited to fix all Scala.js style related issues :) It can happen, for realz
